[ML] Support Gemini thinking budget in inference API

DonalEvans opened the pull request:
Adding support for configuring the thinkingBudget for Gemini 2.5 models when creating chat completion inference endpoints. The thinking_budget field is nested inside the thinking_config object in service_settings.

- Added ThinkingConfig class to contain the thinking_budget field. This results in a less flat structure for the PUT _inference/chat_completion/ call but will make adding support for include_thoughts easier in future
- Added extractOptionalInteger() method to ServiceUtils
- Unit tests for ThinkingConfig class
- Updated existing tests to account for the new object and field

These changes enable elastic/kibana#227590 to be completed

Specification PR: elastic/elasticsearch-specification#5257

Example usage:
```
PUT _inference/chat_completion/my_chat_completion_with_thinking_budget
{
  "service": "googlevertexai",
  "service_settings": {
    "service_account_json": <service account info>,
    "model_id": "gemini-2.5-pro",
    "location": "us-central1",
    "project_id": <project id>
  },
  "task_settings" : {
    "thinking_config": {
      "thinking_budget": 256
    }
  }
}
```

DonalEvans committed:
Support Gemini thinking budget in inference API

Adding support for configuring the thinkingBudget for Gemini 2.5 models
when creating chat completion inference endpoints. The thinking_budget
field is nested inside the thinking_config object in service_settings.

- Added ThinkingConfig class to contain the thinking_budget field. This
  results in a less flat structure for the PUT
  _inference/chat_completion/ call but will make adding support for
  include_thoughts easier in future
- Added extractOptionalInteger() method to ServiceUtils
- Unit tests for ThinkingConfig class
- Updated existing tests to account for the new object and field

These changes enable elastic/kibana#227590 to be completed

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

jonathan-buttner commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/googlevertexai/request/GoogleVertexAiUnifiedChatCompletionRequestEntityTests.java:
```diff
@@ -30,6 +31,8 @@ public class GoogleVertexAiUnifiedChatCompletionRequestEntityTests extends ESTes
 
     private static final String USER_ROLE = "user";
     private static final String ASSISTANT_ROLE = "assistant";
+    private final ThinkingConfig thinkingConfig = new ThinkingConfig(256);
```
nit (optional): I think we could make these static?

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java:
```diff
@@ -780,6 +780,25 @@ public static Integer extractOptionalPositiveInteger(
+        return extractOptionalInteger(map, settingName, scope, validationException, true);
+    }
+
+    public static Integer extractOptionalInteger(
```
Could you add a few tests for this new method?

jonathan-buttner reviewed (commented):
Great work! I left a few comments.

DonalEvans committed:
Apply review feedback

- Move transport version checks from ThinkingConfig to
  GoogleVertexAiChatCompletionServiceSettings
- Remove default value argument from ThinkingConfig.of()
- Add test coverage for ServiceUtils.extractOptionalPositiveInteger()
  and extractOptionalInteger()

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Fix transport versions

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Update docs/changelog/133599.yaml

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

jonathan-buttner commented:
Just wanted to capture our discussion earlier this week. After thinking about this more I think it probably makes sense to move the thinking budget settings to the `task_settings` instead of the `service_settings`. This gives us more flexibility in the future because the user could in theory override the `task_settings` on a per request basis. Because this is for chat completion which adheres to a strict schema format we won't be able to allow individual requests to override the task settings but they can at least set them during the inference endpoint creation request.

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Move thinking config settings into task settings

DonalEvans committed:
Remove unnecessary override from test

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/googlevertexai/completion/GoogleVertexAiChatCompletionTaskSettings.java:
```diff
@@ -0,0 +1,129 @@
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        if (out.getTransportVersion().onOrAfter(TransportVersions.GEMINI_THINKING_BUDGET_ADDED)) {
```
Since we're introducing this class we don't need to check the version here. If we wanted to protect against a mixed cluster scenario we'd need to do this check when we read the named writeable but we haven't been handling that scenario because it's short lived.

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/googlevertexai/completion/GoogleVertexAiChatCompletionTaskSettings.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    public GoogleVertexAiChatCompletionTaskSettings(StreamInput in) throws IOException {
+        if (in.getTransportVersion().onOrAfter(TransportVersions.GEMINI_THINKING_BUDGET_ADDED)) {
```
Let's remove the if block and just have `thinkingConfig = new ThinkingConfig(in);`

jonathan-buttner reviewed (commented):
Looking great, just a few comments.

DonalEvans committed:
Remove transport version checks for task settings class

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

jonathan-buttner reviewed (approved):
Great work!

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget

DonalEvans committed:
Merge branch 'main' into add-gemini-thinking-budget
